pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -1044,6 +1044,210 @@ spec:
+        value: $(tasks.task1.results.result2)
+      taskSpec:
+        params:
+        - name: param1
+          type: string
+        - name: param2
+          type: string
+        steps:
+        - name: foo
+          image: busybox
+          script: 'exit 0'
+  conditions:
+  - message: "Could not find result with name result2 for task task1"
+    reason: InvalidTaskResultReference
+    status: "False"
+    type: Succeeded
+  childReferences:
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: test-pipeline-missing-results-task1
+    pipelineTaskName: task1
+  provenance:
+    featureFlags:
+      RunningInEnvWithInjectedSidecars: true
+      EnableAPIFields: "stable"
+      AwaitSidecarReadiness: true
+      VerificationNoMatchPolicy: "ignore"
+      EnableProvenanceInStatus: true
+      ResultExtractionMethod: "termination-message"
+      MaxResultSize: 4096
+`)
+	d := test.Data{
+		PipelineRuns: prs,
+		TaskRuns:     trs,
+	}
+	prt := newPipelineRunTest(t, d)
+	defer prt.Cancel()
+
+	reconciledRun, clients := prt.reconcileRun("foo", "test-pipeline-missing-results", []string{}, true)
+	if reconciledRun.Status.CompletionTime == nil {
+		t.Errorf("Expected a CompletionTime on invalid PipelineRun but was nil")
+	}
+
+	// The PipelineRun should be marked as InvalidTaskResultReference.
+	if reconciledRun.Status.GetCondition(apis.ConditionSucceeded).Reason != "InvalidTaskResultReference" {
+		t.Errorf("Expected PipelineRun to be InvalidTaskResultReference but condition reason is %s", reconciledRun.Status.GetCondition(apis.ConditionSucceeded).Reason)
+	}
+	if d := cmp.Diff(reconciledRun, expectedPipelineRun, ignoreResourceVersion, ignoreLastTransitionTime, ignoreTypeMeta, ignoreStartTime, ignoreCompletionTime); d != "" {
+		t.Errorf("Expected to see PipelineRun run marked as failed. Diff %s", diff.PrintWantGot(d))
+	}
+
+	// Check that the expected TaskRun was created
+	taskrunList, err := clients.Pipeline.TektonV1beta1().TaskRuns("foo").List(prt.TestAssets.Ctx, metav1.ListOptions{
+		LabelSelector: "tekton.dev/pipelineRun=test-pipeline-missing-results",
+		Limit:         1,
+	})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prs[0].Name, err)
+	}
+	if len(taskrunList.Items) != 1 {
+		t.Fatalf("The pipelineRun should have exactly 1 taskRun for the first task \"task1\", but has %d TaskRuns", len(taskrunList.Items))
+	}
+	taskrunItem := taskrunList.Items[0]
+	if taskrunItem.Labels["tekton.dev/pipelineTask"] != "task1" {
+		t.Fatalf("TaskRun was not found for the task \"task1\"")
+	}
+	if len(taskrunItem.Status.TaskRunResults) != 1 {
+		t.Fatalf("task1 should have produced a result before failing the step")
+	}
+	for _, r := range taskrunItem.Status.TaskRunResults {
+		if r.Name == "result1" && r.Value.StringVal != "123" {
+			t.Fatalf("task1 should have initialized a result \"result1\" to \"123\", but has a result name \"%s\" with the result value \"%s\"", r.Name, r.Value.StringVal)
+		}
+	}
```

Comment:
since results were passed in as input, it doesn't make sense to verify them in the output. I think it would make sense to end the test after verifying the number of taskruns.